runtime/pkg/logbuffer/buffer.go

```diff
@@ -27,35 +26,30 @@ func NewBuffer(maxMessageCount int, maxBufferSize int64) *Buffer {
 	}
 }
 
-func (b *Buffer) AddZapEntry(entry zapcore.Entry, coreFields, entryFields []zapcore.Field) error {
-	size := 0
-	attrs := make(map[string]any, len(coreFields)+len(entryFields))
-	for _, field := range coreFields {
-		attrs[field.Key] = field.String
-		size += len(field.Key) + len(field.String) // approx size
-	}
-
-	for _, field := range entryFields {
-		attrs[field.Key] = field.String
-		size += len(field.Key) + len(field.String) // approx size
+func (b *Buffer) AddZapEntry(entry zapcore.Entry, coreFields, entryFields []zapcore.Field, enc zapcore.Encoder) error {
+	// encode fields using zapcore.Encoder, append to entryFields since coreFields will be empty in many cases
+	// until logger is created `With` method.
+	entryFields = append(entryFields, coreFields...)
+	// send empty entry as we want to store message separately
+	fieldsBuf, err := enc.EncodeEntry(zapcore.Entry{}, entryFields)
+	if err != nil {
+		return err
 	}
+	defer fieldsBuf.Free()
+	payload := fieldsBuf.String()
```

Comment:
It seems odd to pass the Zap encoder along here. Maybe we should move this Zap-specific logic to `BufferedZapCore`. Then this function could be refactored to something like `AddEntry(lvl runtimev1.LogLevel, t time.Time, msg string, payload string) error`.